content/graph/HLD.h

```diff
@@ -1,112 +1,60 @@
 /**
- * Author: Johan Sannemo
- * Date: 2015-02-10
+ * Author: Benjamin Qi, Oleksandr Kulkov, chilli
+ * Date: 2020-01-12
  * License: CC0
- * Source: folkore
+ * Source: https://codeforces.com/blog/entry/53170, https://github.com/bqi343/USACO/blob/master/Implementations/content/graphs%20(12)/Trees%20(10)/HLD%20(10.3).h
  * Description: Decomposes a tree into vertex disjoint heavy paths and light
- *  edges such that the path from any leaf to the root contains at most log(n)
- *  light edges. The function of the HLD can be changed by modifying T, LOW and
- *  f. f is assumed to be associative and commutative.
- * Usage:
- *  HLD hld(G);
- *  hld.update(index, value);
- *  tie(value, lca) = hld.query(n1, n2);
- * Status: Tested at SPOJ
+ * edges such that the path from any leaf to the root contains at most log(n)
+ * light edges. Supports any segtree modifications/queries on paths and
+ * subtrees. Takes as input the full adjacency list.
+ * Status: Will do in morning
  */
 #pragma once
 
-#include "../data-structures/SegmentTree.h"
-
-typedef vector<pii> vpi;
-
-struct Node {
-	int d, par, val, chain = -1, pos = -1;
-};
-
-struct Chain {
-	int par, val;
-	vector<int> nodes;
-	Tree tree;
-};
-
-struct HLD {
-	typedef int T;
-	const T LOW = -(1<<29);
-	void f(T& a, T b) { a = max(a, b); }
-
-	vector<Node> V;
-	vector<Chain> C;
-
-	HLD(vector<vpi>& g) : V(sz(g)) {
-		dfs(0, -1, g, 0);
-		trav(c, C) {
-			c.tree = {sz(c.nodes), 0};
-			for (int ni : c.nodes)
-				c.tree.update(V[ni].pos, V[ni].val);
+#include "../data-structures/LazySegmentTree.h"
+
+template <bool USE_EDGES> struct HLD {
```

Comment:
what does USE_EDGES mean? that nodes have the weight of the edge to their parent?

Reply:
It's whether the values are stored in the edges or in the nodes.

ie: if you have a path (0-1-2) and you modify it, are you modifying 3 values or 2?

The original name was `VALS_IN_EDGES` which I thought was a bit too long - I suppose it might be a more descriptive name.

Regardless, I should definitely add it to the description.

Reply:
I think the longer name is probably better (or a shorter + comment in the description), but I still don't totally get it: where concretely are the edge values stored if it's set? I'm speculating each node contains the weight of the edge to its parent.